cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -50,34 +50,27 @@ extern "C" {
 #define SetGPIO(a) do { GPOS = a; } while (0)
 #define ClearGPIO(a) do { GPOC = a; } while (0)
 
+enum class ExpiryState : uint32_t {OFF = 0, ON = 1, UPDATE = 2}; // for UPDATE, the NMI computes the exact expiry cycle and transitions to ON
+
 // Waveform generator can create tones, PWM, and servos
 typedef struct {
-  uint32_t nextServiceCycle;   // ESP cycle timer when a transition required
-  uint32_t expiryCycle;        // For time-limited waveform, the cycle when this waveform must stop
-  uint32_t nextTimeHighCycles; // Copy over low->high to keep smooth waveform
-  uint32_t nextTimeLowCycles;  // Copy over high->low to keep smooth waveform
+  uint32_t nextServiceCycle;           // ESP cycle timer when a transition required
+  volatile int32_t nextTimeHighCycles; // Copy over low->high to keep smooth waveform
+  volatile int32_t nextTimeLowCycles;  // Copy over high->low to keep smooth waveform
+  volatile uint32_t expiryCycle;       // For time-limited waveform, the cycle when this waveform must stop. In ExpiryState::UPDATE, temporarily holds relative cycle count
+  volatile ExpiryState hasExpiry;      // OFF: expiryCycle (temporarily) ignored. UPDATE: expiryCycle is zero-based, NMI will recompute 
 } Waveform;
 
-static Waveform waveform[17];        // State of all possible pins
```

Comment:
Renaming things like this just makes PRs more complicated to review.  Please try and limit doing this so we can easily see the logic changes vs. things which aren't material.

Reply:
Answered below, actually it's much better reading overall now, I've changed everything to tell the different meanings of cycle apart, and use plural and singular with consideration.